operator/README.md

```diff
@@ -1,2 +1,92 @@
+
+Kubernetes operator for [PMEM-CSI](https://github.com/intel/pmem-csi) driver deployment. Using this operator one can easily deploy and manage PMEM-CSI driver in a Kubernetes cluster.
+
+The PMEM-CSI operator is based on the CoreOS [operator-sdk](https://github.com/operator-framework/operator-sdk) tools and APIs.
+
+## PMEM-CSI Deployment custom resource definition(CRD)
+
+Current PMEM-CSI Deployment object supports below API:
+
+### Deployment
+
+|Field | type | Description |
+|---|---|---|
+| apiVersion | string  | `pmem-csi.intel.com/v1alpha1`|
+| kind | string | `Deployment`|
+| metaData | [ObjectMeta](https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status) | Standard objects metadata. |
+| spec | [DeploymentSpec](#deployment-spec) | Specification of the desired behavior of the deployment |
+
+### DeployemtSpec
+
+|Field | type | Description | Default Value |
+|---|---|---|---|
+| image | ImageInfo | PMEM-CSI driver container image information | `intel/pmem-csi-driver:canary` |
+| logLevel | integer | PMEM-CSI driver logging level | 3 |
+| controller | CotrollerDriver | configuration of PMEM-CSI driver running as controller | |
+| node | NodeDriver | configuration of PMEM-CSI driver running at compute node | |
+
+### ControllerDriver
+
+|Field | type | Description | Default Value |
+|---|---|---|---|
+| provisionerImager | ImageInfo | CSI provisioner container image information | `quay.io/k8scsi/csi-provisioner:v1.2.1` |
+| resources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | describes the compute resource requirements for controller pod |
+
+### NodeDriver
+
+|Field | type | Description | Default Value |
+|---|---|---|---|
+| deviceMode | string | Device management mode to use. Supports one of `lvm` or `direct` | `lvm`
+| registrarImage | ImageInfo | CSI provisioner container image information | `quay.io/k8scsi/csi-node-driver-registrar:v1.1.0` |
+| resources | [ResourceList](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | |
+
+### ImageInfo
+
+|Field | type | Description | Default value |
+|---|---|--|---|
+| registry | string | Docker image registry |
+| name | string | Container image name |
+| tag | string | Container image tag |
+| pullPolicy | string | Image pull policy. either one of `Always`, `Never`, `IfNotPresent` | `IfNotPresent` |
+
+## Pre-requisites
+
+### Build operator image
+
+Build PMEM-CSI operator image using the provided [Dockerfile](https://github.com/intel/pmem-csi/tree/devel/operator/build/Dockerfile) and push the image to local docker registry.
+
+```sh
+docker build --pull -t pmem-csi-operator:canary  --label revision=$(VERSION) -f operator/build/Dockerfile .
+```
+
+### Deploy PMEM-CSI Deployment CRD
+
+```sh
+kubectl create -f https://raw.githubusercontent.com/avalluri/pmem-CSI/operator/operator/deploy/crds/pmem-csi.intel.com_deployments_crd.yaml
```

Comment:
One general question: how useful is it to have the "operator" directory? Do we need it (for example, because the SDK expects it) or can we get rid of it and move the content into the root?

I find it confusing that we now have duplicated "pkg" and "deploy" directories. I'd prefer to have "pkg/operator" (if needed) instead of "operator/pkg".

Reply:
I am not sure though but I believe it's not so easy/clean to regenerate the operator code if we alter the directory structure.

Reply:
Which code needs to be (re)generated? Let's test whether we can do that with everything at the top level.

Reply:
> Which code needs to be (re)generated? Let's test whether we can do that with everything at the top level.

I was referring to the auto-generation of deep copy(zz_generated_deepcoy.go) code. But anyway I gave it a try to meld operator code to pme-csi directory tree in my [branch](https://github.com/avalluri/pmem-CSI/blob/operator-direcotry-tree). Could resolve this issue: https://github.com/avalluri/pmem-CSI/blob/operator-direcotry-tree/operator/operator.make#L15-L24

Reply:
https://github.com/avalluri/pmem-CSI/blob/operator-direcotry-tree/operator/operator.make#L15-L24 is even more complicated than before.

Why can't we just run `operator-sdk generate k8s` directly in the `pmem-csi` directory and put all files where it expects them?

`pkg/apis` at the top level looks good to me. We may end up with more than just the operator API there, but that's fine.

`cmd/manager` is less nice. Why does code generation care about the code that uses the generated code? As far as I can tell, `cmd/manager/main.go` was initially bootstrapped with the SDK, but now is entirely maintained by hand. Correct?

Reply:
> `cmd/manager` is less nice. Why does code generation care about the code that uses the generated code?

It tries to find if the operator is Go operator or not by checking `cmd/manager/main.go`.
The code is here: https://github.com/operator-framework/operator-sdk/blob/6b0412be9d0bb3b3d715c87ddb3afbb2b4f9080e/internal/util/projutil/project_util.go#L93

Reply:
And it's hard-coded:  https://github.com/operator-framework/operator-sdk/blob/6b0412be9d0bb3b3d715c87ddb3afbb2b4f9080e/internal/util/projutil/project_util.go#L39

That's unfortunate and probably worth a feature request to allow reconfiguring that. But for now we can either use `cmd/manager` (as expected) or make it a permanent symlink to `cmd/pmem-csi-operator`.

Reply:
> That's unfortunate and probably worth a feature request to allow reconfiguring that. But for now we can either use `cmd/manager` (as expected) or make it a permanent symlink to `cmd/pmem-csi-operator`.

This not ends here, operatro-sdk is more strict about directory structure :(
```
make operator-generate-k8s 
./_work/bin/operator-sdk-v0.12.0 generate k8s
FATA[0000] must run command in project root dir: project structure requires build/Dockerfile 
operator/operator.make:16: recipe for target 'operator-generate-k8s' failed
make: *** [operator-generate-k8s] Error 1
```

Reply:
This begs the question: is operator SDK really the right tool for us?

If all that we needs is code generation, can't we use the underlying tooling (https://github.com/kubernetes/code-generator, if I'm not mistaken) directly or (if that is too low-level) https://github.com/kubernetes-sigs/kubebuilder instead?